lib/wit/gitrepo.py

```diff
@@ -32,44 +41,46 @@ class GitRepo:
     """
     PKG_DEPENDENCY_FILE = "wit-manifest.json"
 
-    def __init__(self, source, revision, name=None, wsroot=None):
-        self.wsroot = wsroot
-        self.source = source
-        self.revision = revision
-        if name is None:
-            self.name = GitRepo.path_to_name(source)
-        else:
-            self.name = name
+    def __init__(self, name, wsroot: Path):
+        self.name = name
+        self.path = wsroot / name
 
-    def short_revision(self):
-        return self.revision[:8]
+    def download(self, source):
+        if not GitRepo.is_git_repo(self.path):
+            self.clone(source)
+        self.fetch(source)
 
-    def get_path(self):
-        try:
-            return self.path
-
-        except AttributeError:
-            return self.wsroot / self.name
-
-    def set_wsroot(self, wsroot):
-        self.wsroot = wsroot
-
-    def download(self):
-        if not GitRepo.is_git_repo(self.get_path()):
-            self.clone()
-        self.fetch()
-
-    def clone(self):
-        assert not GitRepo.is_git_repo(self.get_path()), \
+    def clone(self, source):
+        assert not GitRepo.is_git_repo(self.path), \
             "Trying to clone and checkout into existing git repo!"
         log.info('Cloning {}...'.format(self.name))
 
-        path = self.get_path()
-        path.mkdir()
-        proc = self._git_command("clone", "--no-checkout", str(self.source), str(path))
-        self._git_check(proc)
+        self.path.mkdir()
+        proc = self._git_command("clone", "--no-checkout", source, str(self.path))
+        try:
+            self._git_check(proc)
+        except GitError:
+            if proc.stderr.startswith("fatal: repository") and \
+               proc.stderr.endswith("does not exist\n"):
+                raise BadSource(self.name, source)
+            else:
+                raise
```

Comment:
Does this `raise` mean it will reraise the same error as before or does it replace it with a generic, information-less exception?

Reply:
It reraises the same exception, and this is probably the more idiomatic form because in Python 2.x, this would be the easiest way of ensuring that the stack trace is preserved as well.